pkg/api/validate/dynamic/sku.go

```diff
@@ -0,0 +1,56 @@
+package dynamic
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/computeskus"
+)
+
+// ValidateWorkerSku uses resourceSkusClient to ensure that the VM sizes listed in the cluster document are available for use in the target region.
+func (dv *dynamic) ValidateVMSku(ctx context.Context, location string, subscriptionID string, oc *api.OpenShiftCluster) error {
+	// Get a list of available worker SKUs, filtering by location. We initialize a new resourceSkusClient here instead of using the one in dv.env,
+	// so that we can determine SKU availability within target cluster subscription instead of within RP subscription.
+	filter := fmt.Sprintf("location eq %s", location)
+	skus, err := dv.resourceSkusClient.List(ctx, filter)
+	if err != nil {
+		return err
+	}
+
+	filteredSkus := computeskus.FilterVMSizes(skus, location)
+	masterProfileSku := string(oc.Properties.MasterProfile.VMSize)
+
+	// Ensure sku in MasterProfile exists in target region
+	if filteredSkus[masterProfileSku] == nil {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "properties.masterProfile.VMSize", "The selected SKU '%v' is unavailable in region '%v'", masterProfileSku, location)
+	}
+
+	// Fail if sku is available, but restricted within the subscription. Restrictions are subscription-specific.
+	// https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/error-sku-not-available
+	isRestricted := computeskus.IsRestricted(filteredSkus, location, masterProfileSku)
+	if isRestricted {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "properties.masterProfile.VMSize", "The selected SKU '%v' is restricted in region '%v' for selected subscription", masterProfileSku, location)
+	}
+
+	// In case there are multiple WorkerProfiles listed in the cluster document (such as post-install),
+	// compare VMSize in each WorkerProfile to the resourceSkusClient call above to ensure that the sku is available in region.
+	for i, workerprofile := range oc.Properties.WorkerProfiles {
+		workerProfileSku := string(workerprofile.VMSize)
+
+		if filteredSkus[workerProfileSku] == nil {
+			return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, fmt.Sprintf("properties.workerProfiles[%d].VMSize", i), "The selected SKU '%v' is unavailable in region '%v'", workerProfileSku, location)
+		}
+
+		isRestricted := computeskus.IsRestricted(filteredSkus, location, workerProfileSku)
+		if isRestricted {
+			return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, fmt.Sprintf("properties.workerProfiles[%d].VMSize", i), "The selected SKU '%v' is restricted in region '%v' for selected subscription", workerProfileSku, location)
+		}
+	}
+
+	return nil
+}
```

Comment:
Checking `filteredSkus` map and `computeskus.IsRestricted(...)` can be moved into a single function. Further reduces code duplication and makes sure that we apply exactly the same checks to both master and worker profiles.

Something like this?

```suggestion
	filteredSkus := computeskus.FilterVMSizes(skus, location)

	err = validateVMSku(filteredSkus, location, string(oc.Properties.MasterProfile.VMSize), "properties.masterProfile.VMSize")
	if err != nil {
		return err
	}

	for i, workerprofile := range oc.Properties.WorkerProfiles {
		err := validateVMSku(filteredSkus, location, string(workerprofile.VMSize), fmt.Sprintf("properties.workerProfiles[%d].VMSize", i))
		if err != nil {
			return err
		}
	}

	return nil
}

func validateVMSku(skus map[string]*mgmtcompute.ResourceSku, location, sku, path string) error {
	if skus[sku] == nil {
		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, path, "The selected SKU '%v' is unavailable in region '%v'", sku, location)
	}

	if computeskus.IsRestricted(skus, location, sku) {
		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, path, "The selected SKU '%v' is restricted in region '%v' for selected subscription", sku, location)
	}

	return nil
}

```